### PR TITLE
CIF-2144 - Create a React project with a Recommendation component

### PR DESCRIPTION
**Pull request by herzog31**
## Description

* Added new React project at `extensions/product-recs/react-components` which will be published to npm as `@adobe/aem-core-cif-product-recs-extension`.
* Added new `jest-extension` job on CircleCI to test the extension react project.
* Updated `validateQueries.test.js` schema test to test against the extended Adobe Commerce 2.4.2 EE schema _with_ Product Recommendation extension installed.
* This is added to Venia via https://github.com/adobe/aem-cif-guides-venia/pull/153.

Overview of new components, contexts, hooks and functions:

| Element | Purpose |
| --- | --- |
| ProductRecsGallery + ProductCard | Renders up to 5 products from the Adobe Commerce Product Recommendations Service |
| StorefrontInstanceContext | Fetches the storefront instance context from GraphQL and provides it to all components that use the [Recommendations JS SDK](https://git.corp.adobe.com/magento-datalake/recommendations-js-sdk). For better performance, the result will be cached in the session storage. |
| useRecommendations | Hook which encapsulates the [Recommendations JS SDK](https://git.corp.adobe.com/magento-datalake/recommendations-js-sdk) and returns the product data based on the provided settings. Customers can reuse this hook to build their own React components that display product recommendation data. |
| useVisibilityObserver | Hook to observe if a given element is in the viewport and was therefore viewed by the visitor. This is used to trigger MSE events. |
| Portal | Updated component to now take either a selector or DOM element as root for a React portal. |
| PortalPlacer | Component that automatically creates portals for each DOM element of the given selector. It will create an instance of the component that is passed as an argument. All dataset attributes of the DOM element will be passed as props to the React component. PortalPlacer will re-render the subtree if it receives any change events from the AEM Sites editor, so changes done by authors are applied correctly. The component is used in Venia, to spawn a new instance of `ProductRecsGallery` for each placeholder that is rendered by the AEM Product Recommendations server-side component.  |
| createProductPageUrl | Simple function to generate a URL to the product redirect servlet based on a product SKU. |
| usePageType | Hook that returns the type of the current page, based on the presence of product or category components. This is required for both MSE and the [Recommendations JS SDK](https://git.corp.adobe.com/magento-datalake/recommendations-js-sdk). |
| useEditorEvents | Hook that listens to events from the AEM Sites editor and updates an internal state for each event. Every component that uses the hook will automatically re-render on editor events. |

## Related Issue

Depends on #606.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes and the overall coverage did not decrease.
- [X] All unit tests pass on CircleCi.
- [X] I ran all tests locally and they pass.

### CODE REVIEW

**Review by dplaton — APPROVED** [review]
I had only minor comments, so 👍  We should still think of a better name, though...

**Review by laurentiumagureanu — APPROVED** [review]

File: extensions/product-recs/react-components/src/components/ProductRecsGallery/ProductCard.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            detail: [{ sku, quantity: 1, virtual: type === 'virtual' }]
+        });
+        document.dispatchEvent(customEvent);
+
+        mse && mse.publish.recsItemAddToCartClick(unitId, productId);
+    };
+
+    const renderPrice = (prices, currency) => {
+        const { minimum, maximum } = prices;
+        const isRange = !!(Math.round(minimum.final * 100) != Math.round(maximum.final * 100));
```
Comment (Buuhuu) [inline extensions/product-recs/react-components/src/components/ProductRecsGallery/ProductCard.js:45]: Is the `!!` necessary here? the `!=` comparison should already result in a `boolean`. You may want to use a strict comparison event `!==`

File: react-components/src/utils/createProductPageUrl.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    // Provided by StoreConfigExporter
+    const pathname = document.body.dataset.storeRootUrl;
+    if (!pathname) {
+        return null;
+    }
+
+    const extension = pathname.substr(pathname.lastIndexOf('.'));
+    const path = pathname.substr(0, pathname.lastIndexOf('.'));
+
+    url.pathname = `${path}.cifproductredirect${extension}/product/${sku}`;
```
Comment (Buuhuu) [inline react-components/src/utils/createProductPageUrl.js:27]: The `/product/` should not be included anymore. The pattern is `{{page}}.cifproductredirect.html/{{sku}}`.

File: react-components/src/utils/createProductPageUrl.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export const createProductPageUrl = sku => {
+    const url = new URL(window.location);
+
+    // Provided by StoreConfigExporter
+    const pathname = document.body.dataset.storeRootUrl;
+    if (!pathname) {
+        return null;
+    }
+
+    const extension = pathname.substr(pathname.lastIndexOf('.'));
```
Comment (Buuhuu) [inline react-components/src/utils/createProductPageUrl.js:24]: We should fix the extension to `html` as this is also how the servlet is registered.

**Review by Buuhuu — CHANGES_REQUESTED** [review]

**Review by Buuhuu — APPROVED** [review]

**Pull request merged**